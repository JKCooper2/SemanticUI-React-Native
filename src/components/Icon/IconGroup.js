// @flow
import React, { PureComponent } from 'react';
import type { Node } from 'react';
import { View, StyleSheet } from 'react-native';

import _ from 'lodash';

import type { IconGroupPropsType } from '~/types/components/Icon';

let SIZES = {
    mini: 8,
    tiny: 12,
    small: 18,
    medium: 24,
    large: 32,
    big: 48,
    huge: 72,
    massive: 96
};

class IconGroup extends PureComponent<IconGroupPropsType> {
    static defaultProps = {
        size: 'medium',
    };

    render() {
        const { size, children } = this.props;

        let viewSize: number = _.get(SIZES, size, SIZES.medium);

        return (
            <View style={{ width: viewSize, height: viewSize }}>
                {React.Children.map(children, (child: Node) => (
                    <View style={[StyleSheet.absoluteFill, { justifyContent: 'center', alignItems: 'center' }]}>
                        {child}
                    </View>
                ))}
            </View>
        );
    }
}


export default IconGroup;