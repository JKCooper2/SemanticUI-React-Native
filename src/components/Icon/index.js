import Icon from './Icon';
import Group from './IconGroup';

export {
    Icon as default,
    Group
};