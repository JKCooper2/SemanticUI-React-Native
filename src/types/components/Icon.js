// @flow
import type { Node } from 'react';
import type { ColorsType } from '../styles';

export type SizeType = 'mini' | 'tiny' | 'small' | 'medium' | 'large' | 'big' | 'huge' | 'massive';
type FlippedType = '' | 'horizontal' | 'vertical';
type TypeType =
    | 'antDesign'
    | 'entypo'
    | 'evilIcons'
    | 'feather'
    | 'fontAwesome'
    | 'fontAwesome5'
    | 'foundation'
    | 'ionicons'
    | 'materialCommunityIcons'
    | 'materialIcons'
    | 'octicons'
    | 'simpleLineIcons'
    | 'zocial';

export type IconPropsType = {
    name: string,
    type: TypeType,
    color: ColorsType,
    size: SizeType,
    flipped: FlippedType,
    fitted: boolean,
    loading: boolean,
    disabled: boolean,
    circular: boolean,
    bordered: boolean,
    inverted: boolean
};

export type IconGroupPropsType = {
    children: Node,
    color?: ColorsType,
    size: SizeType
};

export type IconLibraryPropsType = {
    name: string,
    size: number,
    style: any,
    color: ColorsType
};
