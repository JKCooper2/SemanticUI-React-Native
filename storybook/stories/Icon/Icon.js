// @flow
import React, { Fragment } from 'react';
import { View, Text } from 'react-native';

import { storiesOf } from '@storybook/react-native';

import Icon from '~/components/Icon';

const iconStories = storiesOf('Icon', module);

iconStories.add('Standard', () => <Icon name="rocket" />);

iconStories.add('Type', () => (
    <Fragment>
        <Icon type="antDesign" name="meh" />
        <Icon type="entypo" name="emoji-flirt" />
        <Icon type="evilIcons" name="like" />
        <Icon type="feather" name="database" />
        <Icon type="fontAwesome" name="eye" />
        <Icon type="fontAwesome5" name="rocket" />
        <Icon type="foundation" name="lightbulb" />
        <Icon type="ionicons" name="ios-bicycle" />
        <Icon type="materialCommunityIcons" name="account-check" />
        <Icon type="materialIcons" name="airport-shuttle" />
        <Icon type="octicons" name="beaker" />
        <Icon type="simpleLineIcons" name="user-following" />
        <Icon type="zocial" name="github" />
    </Fragment>
));

iconStories.add('Color', () => <Icon name="rocket" color="red" />);

iconStories.add('Size', () => (
    <Fragment>
        <Icon name="rocket" size="mini" />
        <Icon name="rocket" size="tiny" />
        <Icon name="rocket" size="small" />
        <Icon name="rocket" size="medium" />
        <Icon name="rocket" size="large" />
        <Icon name="rocket" size="big" />
        <Icon name="rocket" size="huge" />
        <Icon name="rocket" size="massive" />
    </Fragment>
));

iconStories.add('Flipped', () => (
    <Fragment>
        <Icon name="rocket" />
        <Icon name="rocket" flipped="horizontal" />
        <Icon name="rocket" flipped="vertical" />
    </Fragment>
));

iconStories.add('Fitted', () => (
    <Fragment>
        <View style={{ display: 'flex', flexDirection: 'row' }}>
            <Text>IC</Text>
            <Icon name="rocket" />
            <Text>ON</Text>
        </View>
        <View style={{ display: 'flex', flexDirection: 'row' }}>
            <Text>IC</Text>
            <Icon name="rocket" fitted />
            <Text>ON</Text>
        </View>
    </Fragment>
));

iconStories.add('Loading', () => <Icon name="rocket" loading />);

iconStories.add('Disabled', () => <Icon name="rocket" disabled />);

iconStories.add('Circular', () => (
    <Fragment>
        <Icon name="rocket" circular />
        <Icon name="rocket" circular inverted />
        <Icon name="rocket" circular color="teal" />
        <Icon name="rocket" circular color="teal" inverted />
    </Fragment>
));

iconStories.add('Bordered', () => (
    <Fragment>
        <Icon name="rocket" bordered />
        <Icon name="rocket" bordered inverted />
        <Icon name="rocket" bordered color="teal" />
        <Icon name="rocket" bordered color="teal" inverted />
    </Fragment>
));

iconStories.add('Group', () => (
    <Icon.Group size="huge">
        <Icon size="big" name="circle" color="red" />
        <Icon name="user" />
    </Icon.Group>
));

export default iconStories;
