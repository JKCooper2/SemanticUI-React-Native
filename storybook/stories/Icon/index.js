// @flow
import Icon from './Icon';

export {
    Icon
};