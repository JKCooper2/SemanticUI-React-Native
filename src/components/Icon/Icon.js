// @flow
// https://oblador.github.io/react-native-vector-icons/
import React, { PureComponent } from 'react';
import type { ComponentType } from 'react';

import { View, Animated, StyleSheet, Easing } from 'react-native';
import type AnimatedValue from 'react-native/Libraries/Animated/src/nodes/AnimatedValue';

import _ from 'lodash';

import AntDesign from 'react-native-vector-icons/AntDesign';
import Entypo from 'react-native-vector-icons/Entypo';
import EvilIcons from 'react-native-vector-icons/EvilIcons';
import Feather from 'react-native-vector-icons/Feather';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import Foundation from 'react-native-vector-icons/Foundation';
import Ionicons from 'react-native-vector-icons/Ionicons';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
import Octicons from 'react-native-vector-icons/Octicons';
import SimpleLineIcons from 'react-native-vector-icons/SimpleLineIcons';
import Zocial from 'react-native-vector-icons/Zocial';

import { spacing, colors } from '~/styles';

import type { IconPropsType, IconLibraryPropsType } from '~/types/components/Icon';
import type { ColorsType } from '~/types/styles';

import IconGroup from './IconGroup';

const SIZES = {
    mini: 8,
    tiny: 12,
    small: 18,
    medium: 24,
    large: 32,
    big: 48,
    huge: 72,
    massive: 96
};

const TYPE = {
    antDesign: AntDesign,
    entypo: Entypo,
    evilIcons: EvilIcons,
    feather: Feather,
    fontAwesome: FontAwesome,
    fontAwesome5: FontAwesome5,
    foundation: Foundation,
    ionicons: Ionicons,
    materialCommunityIcons: MaterialCommunityIcons,
    materialIcons: MaterialIcons,
    octicons: Octicons,
    simpleLineIcons: SimpleLineIcons,
    zocial: Zocial
};

const getColor = (color: ColorsType, disabled: boolean): string => {
    if (disabled) {
        return colors.colors.lightGrey;
    }

    return _.get(colors.colors, color, colors.colors.grey);
};

class Icon extends PureComponent<IconPropsType> {
    static defaultProps = {
        type: 'fontAwesome5',
        color: colors.colors.grey,
        size: 'medium',
        flipped: '',
        fitted: false,
        loading: false,
        disabled: false,
        circular: false,
        bordered: false,
        inverted: false
    };

    static Group = IconGroup;

    spinValue: AnimatedValue;

    constructor(props: IconPropsType) {
        super(props);

        this.spinValue = new Animated.Value(0);
    }

    componentDidMount() {
        this.spin();
    }

    spin = () => {
        if (this.props.loading) {
            this.spinValue.setValue(0);
            Animated.timing(
                this.spinValue,
                {
                    toValue: 1,
                    duration: 1000,
                    easing: Easing.linear,
                    useNativeDriver: true
                }
            ).start(() => this.spin());
        }
    }

    render() {
        const { name, type, color, size, flipped, fitted, loading, disabled, circular, inverted, bordered } = this.props;

        let propStyles = {};

        if (flipped === 'horizontal') {
            propStyles = { ...propStyles, transform: [{ rotate: '90deg' }] };
        }

        if (flipped === 'vertical') {
            propStyles = { ...propStyles, transform: [{ rotate: '270deg' }] };
        }

        let viewSize: number = _.get(SIZES, size, SIZES.medium);

        const spin = loading && this.spinValue.interpolate({
            inputRange: [0, 1],
            outputRange: ['0deg', '360deg']
        });

        let propColor = getColor(color, disabled);

        let IconType: ComponentType<IconLibraryPropsType> = _.get(TYPE, type, TYPE.fontAwesome5);

        let IconComp = (
            <IconType
                name={name}
                size={viewSize}
                style={{ ...propStyles }}
                color={inverted ? colors.colors.white : propColor }
            />
        );

        let borderStyles = {
            borderStyle: 'solid',
            borderWidth: StyleSheet.hairlineWidth,
            borderColor: inverted ? propColor : colors.colors.lightGrey,
            justifyContent: 'center',
            alignItems: 'center',
            backgroundColor: inverted ? propColor : undefined
        };

        if (circular) {
            borderStyles = {
                ...borderStyles,
                borderRadius: viewSize * 1.5 / 2,
                width: viewSize * 1.5,
                height: viewSize * 1.5,
            };
        }

        if (bordered) {
            borderStyles = {
                ...borderStyles,
                borderRadius: 0,
                width: viewSize * 1.2,
                height: viewSize * 1.2,
            };
        }

        if (!circular && !bordered) {
            borderStyles = {};
        }

        return (
            <View style={{ marginHorizontal: fitted ? 0 : spacing.padding.tiny, width: viewSize, height: viewSize, ...borderStyles }}>
                {loading ? (
                    <Animated.View style={{ transform: [{ rotate: spin }] }}>
                        {IconComp}
                    </Animated.View>
                ) : IconComp
                }
            </View>
        );
    }
}

export default Icon;