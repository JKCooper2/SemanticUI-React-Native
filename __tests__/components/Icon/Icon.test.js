// @flow
import React from 'react';
import { shallow } from 'enzyme';

import Icon from '~/components/Icon';

test('renders correctly', () => {
    const icon = shallow(<Icon name="rocket" />);
    expect(icon).toMatchSnapshot();
});
