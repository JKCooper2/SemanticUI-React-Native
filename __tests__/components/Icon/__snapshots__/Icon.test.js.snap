// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  style={
    Object {
      "height": 24,
      "marginHorizontal": 5,
      "width": 24,
    }
  }
>
  <FA5iconClass
    color="#767676"
    light={false}
    name="rocket"
    size={24}
    solid={false}
    style={Object {}}
  />
</Component>
`;
